staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go

```diff
@@ -168,15 +169,14 @@ func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr admission.Attrib
 		if err != nil {
 			switch err := err.(type) {
 			case *webhookutil.ErrCallingWebhook:
-				if ctx.Err() == context.Canceled {
-					klog.Warningf("Context Canceled when calling webhook %v", hook.Name)
-					return err
+				// Ignore context cancelled from webhook metrics
+				if !errors.Is(err.Reason, context.Canceled) {
+					if !ignoreClientCallFailures {
+						rejected = true
+						admissionmetrics.Metrics.ObserveWebhookRejection(ctx, hook.Name, "admit", string(versionedAttr.Attributes.GetOperation()), admissionmetrics.WebhookRejectionCallingWebhookError, int(err.Status.ErrStatus.Code))
+					}
+					admissionmetrics.Metrics.ObserveWebhook(ctx, hook.Name, time.Since(t), rejected, versionedAttr.Attributes, "admit", int(err.Status.ErrStatus.Code))
```

Comment:
skipping the webhook rejection metric for cancelled requests seems ok, but do we really want to lose visibility to the fact that the webhook was called *at all* and the latency we hit for canceled requests?

I was hoping we could revert back to the state at 5e5b3029f3bbfc93c3569f07ad300a5c6057fc58 and narrowly skip the `ObserveWebhookRejection` and `ObserveWebhookFailOpen`/`addFailedOpenAnnotation` bits when the reason was a context cancel

something like this:

```diff
$ git diff -w 5e5b3029f3bbfc93c3569f07ad300a5c6057fc58 HEAD -- staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
diff --git a/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go b/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
index c1d1ca6ff6b..61b17932df1 100644
--- a/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
+++ b/staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/mutating/dispatcher.go
@@ -20,6 +20,7 @@ package mutating
 
 import (
        "context"
+       "errors"
        "fmt"
        "time"
 
@@ -164,14 +165,18 @@ func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr admission.Attrib
                annotator := newWebhookAnnotator(versionedAttr, round, i, hook.Name, invocation.Webhook.GetConfigurationName())
                changed, err := a.callAttrMutatingHook(ctx, hook, invocation, versionedAttr, annotator, o, round, i)
                ignoreClientCallFailures := hook.FailurePolicy != nil && *hook.FailurePolicy == admissionregistrationv1.Ignore
+
                rejected := false
                if err != nil {
                        switch err := err.(type) {
                        case *webhookutil.ErrCallingWebhook:
                                if !ignoreClientCallFailures {
                                        rejected = true
+                                       // Ignore context cancelled from webhook metrics
+                                       if !errors.Is(err.Reason, context.Canceled) {
                                                admissionmetrics.Metrics.ObserveWebhookRejection(ctx, hook.Name, "admit", string(versionedAttr.Attributes.GetOperation()), admissionmetrics.WebhookRejectionCallingWebhookError, int(err.Status.ErrStatus.Code))
                                        }
+                               }
                                admissionmetrics.Metrics.ObserveWebhook(ctx, hook.Name, time.Since(t), rejected, versionedAttr.Attributes, "admit", int(err.Status.ErrStatus.Code))
                        case *webhookutil.ErrWebhookRejection:
                                rejected = true
@@ -199,9 +204,14 @@ func (a *mutatingDispatcher) Dispatch(ctx context.Context, attr admission.Attrib
 
                if callErr, ok := err.(*webhookutil.ErrCallingWebhook); ok {
                        if ignoreClientCallFailures {
+                               // Ignore context cancelled from webhook metrics
+                               if errors.Is(callErr.Reason, context.Canceled) {
+                                       klog.Warningf("Context canceled when calling webhook %v", hook.Name)
+                               } else {
                                        klog.Warningf("Failed calling webhook, failing open %v: %v", hook.Name, callErr)
                                        admissionmetrics.Metrics.ObserveWebhookFailOpen(ctx, hook.Name, "admit")
                                        annotator.addFailedOpenAnnotation()
+                               }
 
                                utilruntime.HandleError(callErr)
```

Reply:
(same comments apply to the validating dispatcher)

Reply:
Sure that makes sense